### PR TITLE
add custom side validation and index on database #29

### LINKED ISSUE #29: Add Custom Server Side Validation.
You can add custom server side validation to your model. You can find an example here: https://stackoverflow.com/questions/71043332/how-and-where-to-validate-uniqueness-of-attribute-property-in-asp-net-core-mvc-u
image

### PR DESCRIPTION
**Pull request by Pantex9**
add custom side validation and index on database #29

### CODE REVIEW

**Comment by Pantex9** [comment]
I have implemented server-side validation for improved security and reliability. 
Additionally, to enhance user experience, I replaced error page redirections with toaster notifications, providing real-time feedback. 
Also, I optimized the Habit table by adding an index on the Name column to improve query performance.

**Review by chrisjamiecarter — CHANGES_REQUESTED** [review]
Hey @Pantex9,

I appreciate the effort! Some very nice improvements here!

I am going to add some comments to the changed files. Please could you action these.

Also, I love the toasts, the the issue was aiming to add validation.
If you see this page: https://stackoverflow.com/a/71059018
I was expecting to see a Validators folder and a UniqueNameAttribute.cs file created for this issue.

So that on a duplicate Habit Name you would see this:
image

Let me know if you have any issues.

Thanks,
@chrisjamiecarter

File: src/HabitTracker.Domain/Helpers/ResponsePackage.cs [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace HabitTracker.Domain.Helpers;
+public class ResponsePackage
+{
+    #region Constructor
+    public ResponsePackage()
```
Comment (chrisjamiecarter) [inline src/HabitTracker.Domain/Helpers/ResponsePackage.cs:12]: See code style throughout the solution.
There should be whitespace between the regions and methods.
But not between regions.

File: src/HabitTracker.Domain/Helpers/ResponsePackage.cs [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+    {
+
+    }
+    #endregion
+
```
Comment (chrisjamiecarter) [inline src/HabitTracker.Domain/Helpers/ResponsePackage.cs:16]: See above

File: src/HabitTracker.Domain/Helpers/ResponsePackage.cs [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+    #region Properties
+
+    public bool IsSuccess { get; set; }
+    public string Message { get; set; }
+    #endregion
```
Comment (chrisjamiecarter) [inline src/HabitTracker.Domain/Helpers/ResponsePackage.cs:23]: See above

File: src/HabitTracker.Infrastructure/Queries/TableQueries.cs [hunk 4]
```diff
@@ -16,6 +16,7 @@ Id TEXT PRIMARY KEY
             ,Measure TEXT NOT NULL
             ,IsActive INTEGER DEFAULT 1
         );
+        CREATE INDEX IF NOT EXISTS IX_Habit_HabitName ON Habit(Name);
```
Comment (chrisjamiecarter) [inline src/HabitTracker.Infrastructure/Queries/TableQueries.cs]: Small issue. Index should be IX_{Table}_{Column}, so expecting: IX_Habit_Name

**Review by chrisjamiecarter — CHANGES_REQUESTED** [review]
Please see added comments

**Comment by Pantex9** [comment]
Okay, I'll do that

**Comment by Pantex9** [comment]
Hi, I resolved the bugs from your comment. also, I align unique name validation attributes with add and update habits.
So when updating but not changing the name, the name can't fail validation, because this is the same name, not changed, but when we update the habit and change the name, then our system will check name from database.

**Review by chrisjamiecarter — CHANGES_REQUESTED** [review]
Hey @Pantex9 ,

Great work! This looks like it will close the issue.

Just one more update...

I've made the infrastructure classes internal to maintain clean architecture principles. This ensures:
- Encapsulation: Infrastructure is only accessible within its own project, keeping the core business logic isolated.
- Maintainability: It prevents accidental use of infrastructure details outside its layer, making future changes easier.
- Reduced Coupling: The core application depends on abstractions, not concrete implementations, promoting flexibility.

It is important in source control that you only work on the issue at hand and not try refactor / change too much stuff out of scope.

Please revert back to internal and let me know if you have any questions!

Thanks
@chrisjamiecarter

**Comment by Pantex9** [comment]
Hey @chrisjamiecarter, 
sorry about that, I tried something and forgot to revert this.
I revert it now

**Review by chrisjamiecarter — APPROVED** [review]
@Pantex9 

Fantastic work!

I have approved, and will merge the pull request.

closes #30

**Pull request merged**